DOC-13422-Eventing-Memory-Quota

Pallavi-Janardhan opened the pull request:
[DOC-13422](https://jira.issues.couchbase.com/browse/DOC-13422)

Created and added a new document for Eventing Memory Quota and linked it from the Sizing Guidelines document. Also modified the existing Sizing Guidelines doc.  

[Docs preview for Sizing Guidelines](https://preview.docs-test.couchbase.com/docs-server-DOC-13422-Eventing-Memory-Quota/server/current/install/sizing-general.html)

and

[Docs preview for Eventing Memory Quota](https://preview.docs-test.couchbase.com/docs-server-DOC-13422-Eventing-Memory-Quota/server/current/eventing/eventing-memory-quota.html)

[Preview credentials](https://couchbasecloud.atlassian.net/wiki/x/dYF_dQ#Preview-docs-for-internal-review)

**NOTE For Writer**: DO NOT Forget that this PR must be merged with the PR --> (https://github.com/couchbaselabs/docs-devex/pull/555)

[DOC-13422]: https://couchbasecloud.atlassian.net/browse/DOC-13422?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

Pallavi-Janardhan committed:
eventing_memory_quota_first_cut_for_my_reference

Pallavi-Janardhan committed:
eventing_memory_quota_first_cut_for_my_reference

Pallavi-Janardhan committed:
eventing_memory_quota_first_cut_for_my_reference_1

Pallavi-Janardhan committed:
eventing_memory_quota_first_cut_for_my_reference_2_sizing_general

Pallavi-Janardhan committed:
eventing_memory_quota_first_cut_for_my_reference_3_sizing_general

Pallavi-Janardhan committed:
eventing_memory_quota_adding_&_Modifying_sizing_guidelines

abhijpes commented on modules/install/pages/eventing-memory-quota.adoc:
```diff
@@ -0,0 +1,168 @@
+
+The Eventing Service memory quota controls specific aspects of memory management but does not act as an absolute ceiling on memory usage.
+
+The Eventing Service memory quota controls specific aspects of memory management, such as queue sizing and garbage collection. 
```
> The Eventing Service memory quota controls specific aspects of memory management

This statement is repetitive as the statement above it.

Pallavi-Janardhan replied:
Deleted the first statement.

abhijpes commented on modules/install/pages/eventing-memory-quota.adoc:
```diff
@@ -0,0 +1,168 @@
+
+The Eventing Service memory quota controls specific aspects of memory management but does not act as an absolute ceiling on memory usage.
+
+The Eventing Service memory quota controls specific aspects of memory management, such as queue sizing and garbage collection. 
```
> such as queue sizing and garbage collection. 

garbage collection -> frequency of garbage collection.

Pallavi-Janardhan replied:
Done. Added.

abhijpes commented on modules/install/pages/eventing-memory-quota.adoc:
```diff
@@ -0,0 +1,168 @@
+= Eventing Memory Quota 
+
+:description: The Eventing Service memory quota does not enforce a hard memory limit on the Eventing subsystem, including producer and worker processes. 
```
> including producer and worker processes

End-user / customer doesn't know what is the "producer" process.
Please go through this entire docs page and remove all mentions of "producer" processes.

Let's only keep using the term "worker processes" (or simply "workers") as the customer / end user is only aware about that.

Pallavi-Janardhan replied:
Done.

abhijpes commented on modules/install/pages/eventing-memory-quota.adoc:
```diff
@@ -0,0 +1,168 @@
+
+== Overview
+
+The Couchbase Eventing Service memory quota does not enforce a hard memory limit on the entire Eventing subsystem, including producer and worker processes. 
```
> including producer and worker processes

Same comment as above on the use of "producer" process.

Pallavi-Janardhan replied:
Done.

abhijpes commented on modules/install/pages/eventing-memory-quota.adoc:
```diff
@@ -0,0 +1,168 @@
+
+The memory quota serves 2 primary purposes:
+
+* **Producer-to-worker queue sizing**: controls the maximum size of each worker's input queue.
```
> Producer-to-worker queue sizing

Same comment as above on the use of "producer" process.

Just use "Worker queue sizing"

Pallavi-Janardhan replied:
Done.

abhijpes commented on modules/install/pages/eventing-memory-quota.adoc:
```diff
@@ -0,0 +1,168 @@
+
+The per-worker quota directly influences 2 key aspects of Eventing runtime behavior.
+
+=== Producer-to-Worker Queue Sizing
```
Producer-to-Worker Queue Sizing

Same comment as above on the use of "producer" process.

Just use "Worker queue sizing"

Pallavi-Janardhan replied:
Done

abhijpes commented on modules/install/pages/eventing-memory-quota.adoc:
```diff
@@ -0,0 +1,168 @@
+
+=== Producer-to-Worker Queue Sizing
+
+Each worker has a bounded input queue that receives changes to documents from the Database Change Protocol (DCP) stream.
```
"Each worker has a bounded input queue" -> "Each worker maintains a bounded input queue"

Pallavi-Janardhan replied:
Done.

abhijpes commented on modules/install/pages/eventing-memory-quota.adoc:
```diff
@@ -0,0 +1,168 @@
+
+=== Producer-to-Worker Queue Sizing
+
+Each worker has a bounded input queue that receives changes to documents from the Database Change Protocol (DCP) stream.
```
> Database Change Protocol (DCP) stream.

Let's not use the term "Database Change Protocol (DCP)". I don't see it mentioned anywhere in our official documentation.

Instead use a generic term like "Document Mutation stream".

Pallavi-Janardhan replied:
We do have a detailed section about DCP in our docs. So, as discussed, will provide reference links to that section, xref:server:learn:clusters-and-availability/intra-cluster-replication.adoc#database-change-protocol[DCP],  from here.

Pallavi-Janardhan replied:
As discussed, DCP is explained in our docs. So, we will keep it as is and provide link to the xref:server:learn:clusters-and-availability/intra-cluster-replication.adoc#database-change-protocol[DCP] section.

abhijpes commented on modules/install/pages/eventing-memory-quota.adoc:
```diff
@@ -0,0 +1,168 @@
+Each worker has a bounded input queue that receives changes to documents from the Database Change Protocol (DCP) stream.
+
+* The per-worker quota determines the maximum size of this queue.
+* When the queue exceeds this size, Eventing throttles the DCP stream to prevent unbounded memory growth.
```
> Eventing throttles the DCP

Same as above for "DCP".

Pallavi-Janardhan replied:
Have provided a link to DCP from this section.

abhijpes commented on modules/install/pages/eventing-memory-quota.adoc:
```diff
@@ -0,0 +1,168 @@
+
+* GC frees memory occupied by unreachable objects that are no longer in use.
+* GC reclaims memory only from objects that are no longer in use. 
+Memory held by live objects in the JSE heap remains allocated, so total usage may exceed the per-worker quota.
```
"JSE" -> "Javascript runtime" * all occurrences

Pallavi-Janardhan replied:
Changed it to JavaScript runtime

Pallavi-Janardhan committed:
DOC-13422_incorporating_technical_comments

sarahlwelton commented on modules/install/pages/modules/architecture/database-change-protocol.adoc:
Why are we committing an empty file? What is the point of this? I think you've made a mistake on the folder structure in this PR, because this file should be in a `pages` folder if it's meant to be documentation.

Pallavi-Janardhan replied:
Yes, guess it was by mistake! Have removed the empty file and fixed the issue.

sarahlwelton reviewed (changes requested):
Hey @Pallavi-Janardhan , can you please confirm that you're making the changes you want to be making in this PR? I can do a more thorough review once you're sure you're committing the right changes. 

Thanks!

Pallavi-Janardhan committed:
removing a page that was added by mistake

Pallavi-Janardhan committed:
moved the eventing memory quota file to the required section and corrected the xref

Pallavi-Janardhan committed:
added the eventing memory quota file to the nav.adoc file

Pallavi-Janardhan reviewed (commented):
Have placed the newly created "eventing-memory-quota.adoc" doc under the "modules/learn/pages/buckets-memory-and-storage" directory and updated the cross-ref from sizing guidelines doc to this. Also have updated the nav.adoc (navigation doc).

Kindly review and let me know your comments or approve.

Pallavi-Janardhan committed:
Merge branch 'release/8.0' into DOC-13422-Eventing-Memory-Quota

Pallavi-Janardhan commented:
@sarahlwelton and @simon-dew can you please have a look at this and let me know if it is ok? If yes, kindly approve so that I can merge it. 

Thanks

sarahlwelton commented on modules/install/pages/sizing-general.adoc:
```diff
@@ -5,7 +5,7 @@
 {description}
 
-When you plan to deploy a Couchbase Server cluster, the most common and important question that comes up is: how many nodes do I need and what size do they need to be?
+The most common and important question that comes up when you plan to deploy a Couchbase Server cluster is: how many nodes do I need and what size do they need to be?
```
```suggestion
The most common and important questions you need to ask when deploying a new Couchbase Server cluster are how many nodes you need, and what size they need to be. 
```

sarahlwelton commented on modules/install/pages/sizing-general.adoc:
```diff
@@ -24,39 +24,42 @@ There needs to be enough capacity in all areas to support everything the system
-This independent service model allows you to take advantage of _Multi-Dimensional Scaling_, whereby a cluster can be fine-tuned for optimal handling of emergent workload-requirements, on a service-by-service basis.
+Couchbase Services are what allow you to use and maintain your data.
+Deploy, maintain, and provision these services independently of each other. 
+This independent service model allows you to take advantage of `Multi-Dimensional Scaling`, whereby a cluster can be fine-tuned for optimal handling of emergent workload-requirements, on a service-by-service basis.
```
```suggestion
Couchbase Services allow you to access and maintain your data.
You can deploy, maintain, and provision these services independently of each other. 
This independent service model allows you to take advantage of Multi-Dimensional Scaling.

Multi-Dimensional scaling lets you fine-tune your cluster for optimal handling of changing workload requirements for each individual Couchbase Service.
```

sarahlwelton commented on modules/install/pages/sizing-general.adoc:
```diff
@@ -24,39 +24,42 @@ There needs to be enough capacity in all areas to support everything the system
-For example, core Data Service operations can often benefit greatly from the _scale out_ of smaller commodity nodes, whereas low latency operations with the Query Service may see a greater benefit from the _scale up_ of hardware resources on a given node.
+Every service has different demands on hardware resources. 
+Multi-Dimensional Scaling plays an important role when sizing your Couchbase cluster, both pre and post-deployment.
+For example, core Data Service operations can often benefit greatly from the `scale out` of smaller commodity nodes, whereas low latency operations with the Query Service may see a greater benefit from the `scale up` of hardware resources on a given node.
```
```suggestion
Every Service has different demands on hardware resources. 
Multi-Dimensional Scaling plays an important role when sizing your Couchbase cluster, both pre and post-deployment.
For example, core Data Service operations can often benefit from scaling out smaller commodity nodes. 
Low latency operations with the Query Service might see a greater benefit from scaling up hardware resources on a given node.
```

sarahlwelton commented on modules/install/pages/sizing-general.adoc:
```diff
@@ -24,39 +24,42 @@ There needs to be enough capacity in all areas to support everything the system
 
 CPU::
-CPU refers to the number of cores and clock speed that are required to run your workload.
+CPU specifies the number of cores and clock speed required to run your workload.
```
```suggestion
CPU controls the number of cores and the clock speed required to run your workload.
```

sarahlwelton replied:
Sorry, just realized you left a 👍 on this but didn't actually make the changes? Can you confirm why?

Pallavi-Janardhan replied:
Guess somehow this did not get committed!  Have taken care of it now.

sarahlwelton commented on modules/install/pages/sizing-general.adoc:
```diff
@@ -24,39 +24,42 @@ There needs to be enough capacity in all areas to support everything the system
-RAM is frequently one of the most crucial areas to size correctly.
-Cached documents allow the reads to be served at low latency and consistently high throughput.
+RAM is often the most crucial areas to size accurately.
+Cached documents provide low-latency reads and consistently high throughput.
```
```suggestion
RAM is often the most crucial area to size.
Cached documents provide low-latency reads and consistently high throughput.
```

sarahlwelton commented on modules/install/pages/sizing-general.adoc:
```diff
@@ -24,39 +24,42 @@ There needs to be enough capacity in all areas to support everything the system
 +
-This resource represents the main memory that you allocate to Couchbase Server and must be determined by the following factors:
+This resource represents the main memory you allocate to Couchbase Server. 
+Determine the allocation based on the following factors:
```
```suggestion
Your RAM represents the main memory you allocate to Couchbase Server. 
Determine your allocation based on the following factors:
```

sarahlwelton commented on modules/install/pages/sizing-general.adoc:
```diff
@@ -24,39 +24,42 @@ There needs to be enough capacity in all areas to support everything the system
-* How much latency you expect from KV/indexing/query performance
+* How much free RAM is available beyond OS and other applications?
+* How much data do you want to store in main memory?
+* How much latency you expect from KV/indexing/query performance?
```
```suggestion
* How much free RAM is available beyond your OS and other applications.
* How much data you want to store in main memory.
* How much latency you expect from your Data, Indexing, and Query Service performance.
```

sarahlwelton commented on modules/install/pages/sizing-general.adoc:
```diff
@@ -24,39 +24,42 @@ There needs to be enough capacity in all areas to support everything the system
 --
 ** All index storage types which need sufficient memory quota allocation for proper functioning.
-** Full Text Search (FTS)
+** Full Text Search (FTS).
```
```suggestion
** The Search Service.
```

sarahlwelton commented on modules/install/pages/sizing-general.adoc:
```diff
@@ -76,7 +79,7 @@ Some components that require RAM are:
 
 | Query Service
-| No RAM-allocation is required for this service.
+| This service does not require any RAM allocation.
```
```suggestion
| The Query Service does not require a RAM allocation.
```

Pallavi-Janardhan replied:
Done.

sarahlwelton replied:
This was not the edit I suggested. Can I ask why?

sarahlwelton commented on modules/install/pages/sizing-general.adoc:
```diff
@@ -89,18 +92,18 @@ Storage (disk space)::
-* [.term]_Disk size_ — Refers to the amount of the disk storage space that is needed to hold your entire data set.
-* [.term]_Disk I/O_ — Is a combination of your sustained read/write rate, the need for compacting the database files, and anything else that requires disk access.
+* [.term]`Disk size` — Specifies the disk storage space needed to hold your entire dataset.
+* [.term]`Disk I/O` — Combines your sustained read/write rate, database file compaction, and any other operations that requires disk access. 
```
```suggestion
* *Disk size* — Specifies the disk storage space needed to hold your entire dataset.
* *Disk I/O* — Combines your sustained read/write rate, database file compaction, and any other operations that requires disk access. 
```

sarahlwelton commented on modules/install/pages/sizing-general.adoc:
```diff
@@ -89,18 +92,18 @@ Storage (disk space)::
-An SSD will give much better performance than a Hard Disk Drive (HDD) when it comes to disk throughput and latency.
+Monitor the fragmentation ratio in the Couchbase user interfaces and trigger compaction processes when needed.
+* Prefer Solid State Drives(SSD), but is not required.
+An SSD gives much better performance than a Hard Disk Drive(HDD) when it comes to disk throughput and latency.
```
```suggestion
Monitor your cluster's fragmentation ratio in the Couchbase Server Web Console and trigger compaction processes as needed.
* Couchbase recommends using Solid State Drives (SSD) when possible.
An SSD gives much better performance than a Hard Disk Drive (HDD) when it comes to disk throughput and latency.
```

sarahlwelton commented on modules/install/pages/sizing-general.adoc:
```diff
@@ -111,23 +114,20 @@ Most deployments can achieve optimal performance with 1 Gbps interconnects, but
+Data Service nodes handle data service operations, such as create/read/update/delete(CRUD).
+The sizing information provided below applies to both the `Couchstore` and `Magma` storage engines.
+Review the differences between these storage engines before attempting sizing.
+For information, see xref:learn:buckets-memory-and-storage/storage-engines.adoc[Storage Engines].
```
```suggestion
Data Service nodes handle data service operations, such as create/read/update/delete (CRUD).
The following sizing information applies to both the Couchstore and Magma storage engines.

Couchbase recommends reviewing the differences between the available storage engines before attempting to size the Data Service nodes in your cluster.
For information, see xref:learn:buckets-memory-and-storage/storage-engines.adoc[].
```

sarahlwelton commented on modules/install/pages/sizing-general.adoc:
```diff
@@ -111,23 +114,20 @@ Most deployments can achieve optimal performance with 1 Gbps interconnects, but
-On the other hand, if your application requires only 10% of data in memory, you need disks with enough space to store all of the data.
-Their read/write rate must also be fast enough to meet your performance goals.
-
-=== RAM Sizing for Data Service Nodes
```
Why did you delete this heading?

Pallavi-Janardhan replied:
I wanted to change the heading level and add another = but guess deleted the whole line by mistake! Apologies. Have added it back.

sarahlwelton commented on modules/install/pages/sizing-general.adoc:
```diff
@@ -111,23 +114,20 @@ Most deployments can achieve optimal performance with 1 Gbps interconnects, but
-
-=== RAM Sizing for Data Service Nodes
+For example, if your working set needs to be fully in-memory, you might need large RAM size.
+On the other hand, if your application requires only 10% of data in-memory, you'll need disks with enough space to store all of the data, and that are fast enough for your read/write operations.
```
```suggestion
For example, if your working set needs to be fully in-memory, your cluster might need more RAM.
If your application requires only 10% of data in-memory, you need disks with enough space to store all of the data, and that are fast enough for your read/write operations.
```

sarahlwelton commented on modules/install/pages/sizing-general.adoc:
```diff
@@ -402,27 +402,29 @@ When sizing, you must account for raw CPU overhead when using a high number of b
 You should allocate additional CPU cores for these workloads. 
 
-* xref:manage:monitor/monitor-intro.adoc[Monitoring] is recommended for CPU usage and System Limits.
+* Refer xref:manage:monitor/monitor-intro.adoc[Monitoring] to monitor CPU usage and System Limits.
```
```suggestion
* For more information about monitoring CPU usage and System Limits, see xref:manage:monitor/monitor-intro.adoc[].
```

sarahlwelton commented on modules/install/pages/sizing-general.adoc:
```diff
@@ -402,27 +402,29 @@ When sizing, you must account for raw CPU overhead when using a high number of b
 == Sizing Index Service Nodes
 
-A node running the Index Service must be sized properly to create and maintain secondary indexes and to perform index scan for {sqlpp} queries.
+Size a node running the Index Service properly to create and maintain secondary indexes and perform index scans for {sqlpp} queries.
```
```suggestion
To create and maintain secondary indexes and perform index scans for {sqlpp} queries, you need to size your Index Service nodes.
```

sarahlwelton commented on modules/install/pages/sizing-general.adoc:
```diff
@@ -402,27 +402,29 @@ When sizing, you must account for raw CPU overhead when using a high number of b
+* What is the minimum, maximum, or average value size of the index field?
+* How many indexes do you need?
+* How many documents need to be indexed?
+* What is the working set percentage of index required memory?
```
```suggestion
* What is the length of your document keys?
* Which fields need to be indexed?
* Will you be using simple or compound indexes?
* What is the minimum, maximum, or average value size of the indexed fields?
* How many indexes do you need?
* How many documents need to be indexed?
* What is the working set percentage of index required memory?
```

sarahlwelton commented on modules/install/pages/sizing-general.adoc:
```diff
@@ -402,27 +402,29 @@ When sizing, you must account for raw CPU overhead when using a high number of b
 *The following is an example use-case for sizing RAM for the Index service:*
 
-The following sizing guide can be used to compute the memory requirement for each individual index and can be used to determine the total RAM quota required for the Index service.
+Use the following sizing guide to compute the memory requirement for each individual index and to determine the total RAM quota required for the Index service.
```
```suggestion
Use the following sizing guide to compute the memory requirement for each individual index and to determine the total RAM quota required for the Index Service.
```

sarahlwelton commented on modules/install/pages/sizing-general.adoc:
```diff
@@ -507,7 +509,8 @@ Based on the above formula, these are the suggested sizing guidelines:
 
-The above example shows the memory requirement of a secondary index with 10M index entries, each with 50 bytes size of secondary key and 30 bytes size of documentID. The memory usage requirements are 2.5GB(Nitro, 100% resident), 1GB(plasma, 20% resident), 800MB(Forestdb, 20% resident).
+The above example shows the memory requirement of a secondary index with 10M index entries, each with 50 bytes size of secondary key and 30 bytes size of documentID. 
+The memory usage requirements are 2.5{nbsp}GB(Nitro, 100% resident), 1{nbsp}GB(plasma, 20% resident), 800{nbsp}MB(ForestDB, 20% resident).
```
```suggestion
The memory usage requirements are 2.5{nbsp}GB (Nitro, 100% resident), 1{nbsp}GB (plasma, 20% resident), 800{nbsp}MB (ForestDB, 20% resident).
```

sarahlwelton commented on modules/install/pages/sizing-general.adoc:
```diff
@@ -529,22 +532,22 @@ NOTE: The storage engine used in the sizing calculation corresponds to the stora
 A node that runs the Query Service executes queries for your application needs.
 
-Since the Query Service doesn’t need to persist data to disk, there are very minimal resource requirements for disk space and disk I/O.
+Since the Query Service doesn't need to persist data to disk, there are minimal resource requirements for disk space and disk I/O.
```
```suggestion
Since the Query Service does not need to persist data to disk, there are minimal resource requirements for disk space and disk I/O.
```

sarahlwelton commented on modules/install/pages/sizing-general.adoc:
```diff
@@ -529,22 +532,22 @@ NOTE: The storage engine used in the sizing calculation corresponds to the stora
 You only need to consider CPU and memory.
 
-There are a few questions that will help size the cluster:
+Questions that help in sizing the cluster:
```
```suggestion
Answer the following questions to help size the Query Service nodes on your cluster:
```

sarahlwelton commented on modules/install/pages/sizing-general.adoc:
```diff
@@ -529,22 +532,22 @@ NOTE: The storage engine used in the sizing calculation corresponds to the stora
-. What are the throughput and latency requirements for your queries?
+* What types of queries do you need to run?
+* Do you need to run `stale=ok` or `stale=false` queries?
+* Are the queries simple or complex (requiring JOINs, for example)?
```
```suggestion
* Are the queries simple or complex? For example, do you need to use JOINs?
```

sarahlwelton commented on modules/install/pages/sizing-general.adoc:
```diff
@@ -529,22 +532,22 @@ NOTE: The storage engine used in the sizing calculation corresponds to the stora
-If either limit is reached, additional queries are rejected with a `503` error.
+The formula used to calculate the number of queries that's processed simultaneously is `CPU&#95;cores * 4`.
+The formula used to calculate the maximum queue-length for queries is `CPU&#95;cores * 256`.
+The system rejects additional queries with a 503 error once it reaches the limits.
```
```suggestion
If you reach either limit, the system rejects additional queries with a 503 error.
```

sarahlwelton commented on modules/install/pages/sizing-general.adoc:
```diff
@@ -554,68 +557,95 @@ The Analytics Service is dependent on the Data Service and requires the Data ser
-* Ensure that the data space for Analytics node takes into account metadata replicas. The Analytics Service currently only replicates metadata and not the actual data. There is a small overhead for metadata replicas as metadata is usually small.
+* Make sure that the data space for Analytics node takes into account metadata replicas. 
+The Analytics Service only replicates the metadata and not the actual data. 
+There's a small overhead for metadata replicas as metadata is generally small.
```
```suggestion
* Make sure that the data space for your Analytics Service nodes takes into account metadata replicas. 
The Analytics Service only replicates the metadata and not the actual data. 
There's a small overhead for metadata replicas as metadata is generally small.
```

sarahlwelton commented on modules/install/pages/sizing-general.adoc:
```diff
@@ -554,68 +557,95 @@ The Analytics Service is dependent on the Data Service and requires the Data ser
+The Analytics Service only replicates the metadata and not the actual data. 
+There's a small overhead for metadata replicas as metadata is generally small.
 
-* When evaluating a query, the Analytics engine uses temporary disk space. The type of query being executed can impact the amount of temporary disk space required. For example, a query with heavy JOINs, aggregates, windowing, or more predicates will require more temporary disk space. Typically, the temporary disk space can be 2x the data space.
+* When evaluating a query, the Analytics engine uses temporary disk space. 
+The query type the required amount of temporary disk space. 
+For example, queries with heavy JOINs, aggregates, windowing, or additional predicates require more temporary disk space. 
+Typically, the temporary disk space can be 2x the data space.
```
```suggestion
* When evaluating a query, the Analytics engine uses temporary disk space. 
The type of query you want to run determines the required amount of temporary disk space. 
+
For example, queries with heavy JOINs, aggregates, windowing, or additional predicates require more temporary disk space. 
Typically, the temporary disk space can be 2x the data space.
```

sarahlwelton commented on modules/install/pages/sizing-general.adoc:
```diff
@@ -554,68 +557,95 @@ The Analytics Service is dependent on the Data Service and requires the Data ser
-* When ingesting data from the the Data Service into the Analytics Service a filter can be provided that reduces the size of the data that is ingested and also the storage size for the Analytics Service proportionally.
+* When you load data from the Data Service into the Analytics Service, you can apply a filter to reduce both the loaded data size and the Analytics Service storage requirements proportionally.
 
 === Disk types and partioning
```
```suggestion
=== Disk Types and Partitioning
```

sarahlwelton commented on modules/install/pages/sizing-general.adoc:
```diff
@@ -554,68 +557,95 @@ The Analytics Service is dependent on the Data Service and requires the Data ser
-During query execution, Analytics’s query engine attempts to concurrently read and process data from all data partitions. Because of that, the Input/Output Operations per Second (IOPS) of the actual physical disk in which each data partition resides plays a major role in determining the query execution time.
-Modern storage devices such as SSDs have much higher IOPS and can deal better with concurrent reads than HDDs. Therefore, having a single data partition on devices with high IOPS will not fully utilize their capabilities.
+During query execution, the Analytics query engine concurrently reads and processes data from all partitions. 
+The Input/Output Operations per Second(IOPS) of the physical disk that hosts the data partitions plays a major role in determining the query execution time.
```
```suggestion
The Input/Output Operations per Second (IOPS) of the physical disk that hosts the data partitions plays a major role in determining the query execution time.
```

sarahlwelton commented on modules/install/pages/sizing-general.adoc:
```diff
@@ -554,68 +557,95 @@ The Analytics Service is dependent on the Data Service and requires the Data ser
+During query execution, the Analytics query engine concurrently reads and processes data from all partitions. 
+The Input/Output Operations per Second(IOPS) of the physical disk that hosts the data partitions plays a major role in determining the query execution time.
+Modern storage devices such as SSDs have much higher IOPS and can deal better with concurrent reads than HDDs. 
+Therefore, a single data partition under-utilizes high-IOPS devices.
```
```suggestion
A single data partition underutilizes high IOPS devices.
```

sarahlwelton commented on modules/install/pages/sizing-general.adoc:
```diff
@@ -554,68 +557,95 @@ The Analytics Service is dependent on the Data Service and requires the Data ser
-To simplify the setup of the typical case of a node having a single modern storage device, the Analytics service automatically creates multiple data partitions within the same storage device if and only if a single “Analytics Disk Path” is specified during the node initialization. The number of automatically created data partitions is based on this formula:
+To simplify setup for nodes with a single modern storage device, the Analytics Service creates multiple data partitions on the same storage device.
+It does this only when you specify a single Analytics disk path during node initialization.
+The service determines the number of partitions using the following formula:
```
```suggestion
The Analytics Service determines the number of partitions using the following formula:
```

sarahlwelton commented on modules/install/pages/sizing-general.adoc:
```diff
@@ -554,68 +557,95 @@ The Analytics Service is dependent on the Data Service and requires the Data ser
+For example, if a node has 8 virtual cores and the Analytics Service has at least 8{nbsp}GB of memory, the system creates 8 data partitions on that node.
+Similarly, for a node with 32 virtual cores and 16{nbsp}GB memory, the system creates 16 partitions, the maximum for automatic partitioning.
 
 === Index considerations
```
```suggestion
=== Index Considerations
```

sarahlwelton commented on modules/install/pages/sizing-general.adoc:
```diff
@@ -554,68 +557,95 @@ The Analytics Service is dependent on the Data Service and requires the Data ser
 == Sizing Eventing Service Nodes
 
-Eventing is a compute oriented service. By default, Eventing service has one worker and each worker has two threads of execution. You can scale eventing both vertically by adding more workers or horizontally by adding more nodes. The Eventing service will partition vBuckets across the number of available nodes.
+Eventing is a compute oriented service. 
```
```suggestion
Eventing is a compute-oriented service. 
```

sarahlwelton commented on modules/install/pages/sizing-general.adoc:
```diff
@@ -554,68 +557,95 @@ The Analytics Service is dependent on the Data Service and requires the Data ser
+Eventing is a compute oriented service. 
+By default, Eventing service has 1 worker and each worker has 2 threads of execution. 
+You can scale eventing both vertically by adding more workers or horizontally by adding more nodes. 
+The Eventing service partitions the vBuckets across the number of available nodes.
```
```suggestion
By default, the Eventing Service has 1 worker and each worker has 2 threads of execution. 
You can scale the Eventing Service both vertically by adding more workers or horizontally by adding more nodes. 
The Eventing Service partitions the vBuckets across the number of available nodes.
```

sarahlwelton commented on modules/install/pages/sizing-general.adoc:
```diff
@@ -554,68 +557,95 @@ The Analytics Service is dependent on the Data Service and requires the Data ser
 === CPU
 
-Because Eventing allows arbitrary code, JavaScript, to be written and run, it is difficult to come up with a perfect sizing formula unless all Functions have been designed and their KV ops, Query ops, and cURLops are known along with the expected mutation rate.
+Eventing runs arbitrary JavaScript code.
```
```suggestion
The Eventing Service allows you to write and run arbitrary JavaScript code.
```

sarahlwelton commented on modules/install/pages/sizing-general.adoc:
```diff
@@ -554,68 +557,95 @@ The Analytics Service is dependent on the Data Service and requires the Data ser
 For example, if you process 100K mutations per second and only match 1 out of 1000 patterns, then perform some intense computation on the matched 100 items in your Eventing Function, you need 100X less compute than if you performed the intense computation on each mutation.
 
-Eventing also can perform I/O to external REST endpoints via a synchronous HTTP/S cURL call. In this case, Eventing typically blocks on I/O and doesn’t need much CPU. However. if you want high throughput to overcome bandwidth, you will need more workers and thus more cores.
+Eventing also can perform I/O to external REST endpoints via a synchronous HTTP/S cURL call. 
```
```suggestion
The Eventing Service also can perform I/O to external REST endpoints through a synchronous HTTP/S cURL call. 
```

sarahlwelton commented on modules/install/pages/sizing-general.adoc:
```diff
@@ -554,68 +557,95 @@ The Analytics Service is dependent on the Data Service and requires the Data ser
 
-Eventing also can perform I/O to external REST endpoints via a synchronous HTTP/S cURL call. In this case, Eventing typically blocks on I/O and doesn’t need much CPU. However. if you want high throughput to overcome bandwidth, you will need more workers and thus more cores.
+Eventing also can perform I/O to external REST endpoints via a synchronous HTTP/S cURL call. 
+Eventing typically blocks on I/O and requires little CPU.
```
```suggestion
The Eventing Service typically blocks on I/O and does not need much CPU.
```

sarahlwelton commented on modules/install/pages/sizing-general.adoc:
```diff
@@ -554,68 +557,95 @@ The Analytics Service is dependent on the Data Service and requires the Data ser
+Achieving high throughput to overcome bandwidth requires additional workers and cores. 
 
-8 vCPUs or 4 physical cores should be considered a good start for running a few Eventing Functions.
+Use 8 vCPUs or 4 physical cores to run Eventing Functions.
```
```suggestion
For a starting size for your Eventing Service nodes and running Eventing Functions, use 8 vCPUs or 4 physical cores.
```

sarahlwelton commented on modules/install/pages/sizing-general.adoc:
```diff
@@ -554,68 +557,95 @@ The Analytics Service is dependent on the Data Service and requires the Data ser
 === RAM
 
-In general, the Eventing memory quota of 256 MB is sufficient for almost all workloads.
+The xref:learn:buckets-memory-and-storage/eventing-memory-quota.adoc[Eventing Memory Quota] doc discusses the Eventing Service memory quota in detail.
```
```suggestion
Couchbase recommends a memory quota of 256{nbsp}MB for the Eventing Service. 
For more information about the Eventing Service's memory quota, see xref:learn:buckets-memory-and-storage/eventing-memory-quota.adoc[].
```

sarahlwelton commented on modules/install/pages/sizing-general.adoc:
```diff
@@ -554,68 +557,95 @@ The Analytics Service is dependent on the Data Service and requires the Data ser
+* The distribution of the total memory quota across workers
+* The limitations and considerations
+* Best practices and recommendations
+--
```
```suggestion

```

sarahlwelton commented on modules/install/pages/sizing-general.adoc:
```diff
@@ -554,68 +557,95 @@ The Analytics Service is dependent on the Data Service and requires the Data ser
 === Eventing Storage Collection (previously Metadata Bucket)
 
-Eventing functions store less than 2048 docs per Function. If timers are not used or if you have less than a few thousand active timers, then the size of the Eventing storage collection can simply be in a bucket with a minimum size 100 MB.
+Each Eventing function stores fewer than 2048 documents in its Eventing storage collection. 
+If timers are not used or if the active timers count does not exceed the per-function document limit, store the Eventing storage collection in a 100 MB bucket.
```
```suggestion
Each Eventing Function stores fewer than 2048 documents in its Eventing storage collection. 
If timers are not used or if your number of active timers does not exceed the per-function document limit, store the Eventing storage collection in a 100{nbsp}MB bucket.
```

sarahlwelton commented on modules/install/pages/sizing-general.adoc:
```diff
@@ -554,68 +557,95 @@ The Analytics Service is dependent on the Data Service and requires the Data ser
+Each active timer requires 800 bytes, plus the size of the passed context, which represents the state supplied to the function at future execution.
 
-If you have a context of 200 bytes (total 1K/timer), then for 100,000 active timers you'll need 100 MB of additional space in this bucket.
+A 200-byte context results in 1 KB of storage per active timer.
+100,000 active timers require 100 MB of additional bucket space.
```
```suggestion
A 200-byte context results in 1{nbsp}KB of storage per active timer.
100,000 active timers require 100{nbsp}MB of additional bucket space.
```

sarahlwelton commented on modules/install/pages/sizing-general.adoc:
```diff
@@ -554,68 +557,95 @@ The Analytics Service is dependent on the Data Service and requires the Data ser
+100,000 active timers require 100 MB of additional bucket space.
 
-As a best practice, it's recommended to keep this collection 100% resident, so that it's always available in-memory.
+As a best practice, keep this collection fully resident in-memory to make sure constant availability.
```
```suggestion
As a best practice, keep this collection fully resident in-memory.
```

sarahlwelton commented on modules/install/pages/sizing-general.adoc:
```diff
@@ -554,68 +557,95 @@ The Analytics Service is dependent on the Data Service and requires the Data ser
+As a best practice, keep this collection fully resident in-memory to make sure constant availability.
 
-NOTE: This collection is shared across all your Eventing Functions.
+NOTE: All Eventing functions use this collection.
```
```suggestion
NOTE: All your Eventing Functions share this collection.
```

sarahlwelton commented on modules/install/pages/sizing-general.adoc:
```diff
@@ -631,13 +661,14 @@ The hardware requirements for running a backup cluster are as follows:
-For example, XDCR needs 1-2 additional CPU cores per stream. In some cases, it also requires additional RAM and network resources. If a cluster is not sized to handle _both_ the existing workload _and_ the new XDCR streams, the performance of both XDCR and the cluster overall may be negatively impacted.
+For example, XDCR needs 1-2 additional CPU cores per stream. 
+In some cases, it also requires additional RAM and network resources. 
+If a cluster is not sized to handle `both` the existing workload `and` the new XDCR streams, the performance of both XDCR and the cluster overall may be negatively impacted.
```
```suggestion
If your cluster is not sized to handle both the existing workload and the new XDCR streams, the performance of both XDCR and your cluster might be negatively impacted.
```

sarahlwelton commented on modules/learn/pages/buckets-memory-and-storage/eventing-memory-quota.adoc:
```diff
@@ -0,0 +1,166 @@
+= Eventing Memory Quota 
```
```suggestion
= Eventing Service Memory Quota 
```

Pallavi-Janardhan replied:
Done.

sarahlwelton commented on modules/learn/pages/buckets-memory-and-storage/eventing-memory-quota.adoc:
```diff
@@ -0,0 +1,166 @@
+
+:description: The Eventing Service memory quota does not enforce a hard memory limit on the Eventing subsystem, including worker processes. 
+
+This page explains how the memory quota actually works, how Eventing distributes it across workers, and its limitations in non-containerized environments.
```
```suggestion
This page explains how the memory quota actually works, how the Eventing Service distributes it across workers, and its limitations in non-containerized environments.
```

Pallavi-Janardhan replied:
Done.

sarahlwelton commented on modules/learn/pages/buckets-memory-and-storage/eventing-memory-quota.adoc:
```diff
@@ -0,0 +1,166 @@
+[abstract]
+{description}
+
+== Overview
```
```suggestion
== About the Eventing Service Memory Quota
```

Pallavi-Janardhan replied:
Done.

sarahlwelton commented on modules/learn/pages/buckets-memory-and-storage/eventing-memory-quota.adoc:
```diff
@@ -0,0 +1,166 @@
+
+== Overview
+
+The Couchbase Eventing Service memory quota does not enforce a hard memory limit on the entire Eventing subsystem, including worker processes. 
```
```suggestion

```
You said this already. Decide where on the page you want this information to go instead of repeating it so close on the same page.

sarahlwelton commented on modules/learn/pages/buckets-memory-and-storage/eventing-memory-quota.adoc:
```diff
@@ -0,0 +1,166 @@
+The Couchbase Eventing Service memory quota does not enforce a hard memory limit on the entire Eventing subsystem, including worker processes. 
+The service uses the quota for queue sizing and JavaScript (JS) runtime memory heap, not as an absolute memory cap. 
+In VM or bare-metal deployments, Eventing functions can exceed the configured quota at runtime.
+Understanding this behavior helps you to appropriately size and monitor Eventing functions in production environments.
```
```suggestion
The Eventing Service uses the quota for:

* <<worker-queue-sizing,>>: Controls the maximum size of each worker's input queue.
* <<gc-trigger,>>: Determines when the Eventing Service invokes the JavaScript environment's GC to reclaim memory.

The memory quota is not an absolute memory cap.
The quota does not restrict the total memory consumed by JavaScript runtime heaps or Eventing processes.

The Eventing Service divides its total memory quota uniformly across all workers in all deployed functions:
....
Per-Worker quota = Total memory quota ÷ Total number of workers
....

For example, if you had a memory quota of 256{nbsp}MB, with 4 total workers, each worker would use 64{nbsp}MB of memory, or 256 MB ÷ 4 workers.

Each worker receives an equal share of the total quota, regardless of the actual memory usage of individual functions.

As you deploy more functions or add workers per function, the per-worker quota decreases proportionally.
A smaller per-worker quota can affect performance if individual functions require substantial memory for processing.

In VM or bare-metal deployments, Eventing Functions can exceed the configured quota at runtime.
Understanding this behavior helps you to appropriately size and monitor your Eventing Functions in production environments.
```

sarahlwelton commented on modules/learn/pages/buckets-memory-and-storage/eventing-memory-quota.adoc:
```diff
@@ -0,0 +1,166 @@
+In VM or bare-metal deployments, Eventing functions can exceed the configured quota at runtime.
+Understanding this behavior helps you to appropriately size and monitor Eventing functions in production environments.
+
+== How Memory Quota Works
```
```suggestion

```

sarahlwelton commented on modules/learn/pages/buckets-memory-and-storage/eventing-memory-quota.adoc:
```diff
@@ -0,0 +1,166 @@
+== How Memory Quota Works
+
+The Eventing Service memory quota controls specific aspects of memory management, such as queue sizing and frequency of garbage collection. 
+However, it does not act as an absolute ceiling on memory usage.
```
```suggestion

```

sarahlwelton commented on modules/learn/pages/buckets-memory-and-storage/eventing-memory-quota.adoc:
```diff
@@ -0,0 +1,166 @@
+
+== How Memory Quota Works
+
+The Eventing Service memory quota controls specific aspects of memory management, such as queue sizing and frequency of garbage collection. 
```
```suggestion

```

sarahlwelton commented on modules/learn/pages/buckets-memory-and-storage/eventing-memory-quota.adoc:
```diff
@@ -0,0 +1,166 @@
+=== Minimum Size
+
+The memory quota must be a minimum of 256 MB. 
+Couchbase does not support values lesser than 256 MB.
```
```suggestion
The memory quota for the Eventing Service must be a minimum of 256{nbsp}MB. 
Couchbase does not support memory quota values smaller than 256{nbsp}MB.
```

sarahlwelton commented on modules/learn/pages/buckets-memory-and-storage/eventing-memory-quota.adoc:
```diff
@@ -0,0 +1,166 @@
+
+* **Worker queue sizing**: controls the maximum size of each worker's input queue.
+* **Garbage Collection (GC) triggering**: determines when Eventing invokes JavaScript environment GC to reclaim memory.
+The quota does not restrict the total memory consumed by JavaScript runtime heaps or Eventing processes.
```
```suggestion

```

sarahlwelton commented on modules/learn/pages/buckets-memory-and-storage/eventing-memory-quota.adoc:
```diff
@@ -0,0 +1,166 @@
+
+== Per-Worker Distribution
+
+The Eventing Service divides the total memory quota uniformly across all workers in all deployed functions.
```
```suggestion

```

sarahlwelton commented on modules/learn/pages/buckets-memory-and-storage/eventing-memory-quota.adoc:
```diff
@@ -0,0 +1,166 @@
+* **Garbage Collection (GC) triggering**: determines when Eventing invokes JavaScript environment GC to reclaim memory.
+The quota does not restrict the total memory consumed by JavaScript runtime heaps or Eventing processes.
+
+== Per-Worker Distribution
```
```suggestion

```

sarahlwelton commented on modules/learn/pages/buckets-memory-and-storage/eventing-memory-quota.adoc:
```diff
@@ -0,0 +1,166 @@
+Eventing calculates the per-worker quota as follows:
+....
+Per-Worker quota = Total memory quota ÷ Total number of workers
+....
```
```suggestion

```

sarahlwelton commented on modules/learn/pages/buckets-memory-and-storage/eventing-memory-quota.adoc:
```diff
@@ -0,0 +1,166 @@
+// == Related Links 
+// * xref:eventing-service-overview.adoc[Eventing Service Overview]
+// * xref:eventing-service-settings.adoc[Eventing Service Settings]
+// * 
```
Clean this up, please. Don't just leave this commented out with this kind of comment.

sarahlwelton commented on modules/learn/pages/buckets-memory-and-storage/eventing-memory-quota.adoc:
```diff
@@ -0,0 +1,166 @@
+* Total workers (across all deployed functions): 4
+* Per-worker budget: 64 MB (256 MB ÷ 4 workers)
+
+Each worker receives an equal share of the total quota, regardless of the actual memory usage of individual functions.
```
```suggestion

```

sarahlwelton commented on modules/ROOT/nav.adoc:
```diff
@@ -51,6 +51,7 @@ include::third-party:partial$nav.adoc[]
   *** xref:learn:buckets-memory-and-storage/compression.adoc[Compression]
  ** xref:learn:buckets-memory-and-storage/memory-and-storage.adoc[Memory and Storage]
   *** xref:learn:buckets-memory-and-storage/memory.adoc[Memory]
+  *** xref:learn:buckets-memory-and-storage/eventing-memory-quota.adoc[Eventing Memory Quota]
```
Why are we not putting this together with the rest of the Eventing documentation? Why did we decide to put it in the memory and storage section? This doesn't quite make sense.

Pallavi-Janardhan replied:
It was initially decided by the SMEs to put this chapter under Memory and Storage section. After seeing your note, I checked with the SMEs and put this doc under Eventing documentation, between Debugging and Statistics, as per the SME's suggestion.

sarahlwelton commented on modules/learn/pages/buckets-memory-and-storage/eventing-memory-quota.adoc:
```diff
@@ -0,0 +1,166 @@
+=== Impact of Worker Count
+
+As you deploy more functions or add workers per function, the per-worker quota decreases proportionally.
+A smaller per-worker quota can affect performance if individual functions require substantial memory for processing.
```
```suggestion

```

sarahlwelton commented on modules/learn/pages/buckets-memory-and-storage/eventing-memory-quota.adoc:
```diff
@@ -0,0 +1,166 @@
+
+== Queue Sizing and GC Behavior
+
+The per-worker quota directly influences 2 key aspects of Eventing runtime behavior.
```
```suggestion

```

sarahlwelton commented on modules/learn/pages/buckets-memory-and-storage/eventing-memory-quota.adoc:
```diff
@@ -0,0 +1,166 @@
+
+The per-worker quota directly influences 2 key aspects of Eventing runtime behavior.
+
+=== Worker Queue Sizing
```
```suggestion
== Worker Queue Sizing
```

sarahlwelton commented on modules/learn/pages/buckets-memory-and-storage/eventing-memory-quota.adoc:
```diff
@@ -0,0 +1,166 @@
+* When the queue exceeds this size, Eventing throttles the xref:server:learn:clusters-and-availability/intra-cluster-replication.adoc#database-change-protocol[DCP] stream to prevent unbounded memory growth.
+* Throttling temporarily pauses mutation flow, ensuring workers do not become overwhelmed when processing lags behind the mutation rate. 
+
+=== Garbage Collection (GC) Triggering
```
```suggestion
[#gc-trigger]
== Garbage Collection (GC) Triggering
```

sarahlwelton commented on modules/learn/pages/buckets-memory-and-storage/eventing-memory-quota.adoc:
```diff
@@ -0,0 +1,166 @@
+
+=== Worker Queue Sizing
+
+Each worker maintains a bounded input queue that receives changes to documents from the Database Change Protocol (xref:server:learn:clusters-and-availability/intra-cluster-replication.adoc#database-change-protocol[DCP]) stream. 
```
```suggestion
Each worker maintains a bounded input queue that receives changes to documents from the xref:server:learn:clusters-and-availability/intra-cluster-replication.adoc#database-change-protocol[Database Change Protocol (DCP)] stream. 
```

sarahlwelton commented on modules/learn/pages/buckets-memory-and-storage/eventing-memory-quota.adoc:
```diff
@@ -0,0 +1,166 @@
+
+* The per-worker quota determines the maximum size of this queue.
+* When the queue exceeds this size, Eventing throttles the xref:server:learn:clusters-and-availability/intra-cluster-replication.adoc#database-change-protocol[DCP] stream to prevent unbounded memory growth.
+* Throttling temporarily pauses mutation flow, ensuring workers do not become overwhelmed when processing lags behind the mutation rate. 
```
```suggestion
The per-worker quota determines the maximum size of this queue.

When the queue exceeds this size, the Eventing Service throttles the DCP stream to prevent unbounded memory growth.

Throttling temporarily pauses the mutation flow, ensuring workers do not become overwhelmed when processing lags behind the mutation rate. 
```

sarahlwelton commented on modules/learn/pages/buckets-memory-and-storage/eventing-memory-quota.adoc:
```diff
@@ -0,0 +1,166 @@
+NOTE: When memory usage approaches the per-worker quota limit, the JavaScript runtime may trigger garbage collection to reclaim memory. 
+However, the garbage collector may delay the stop-the-world collection because it's optimized for throughput. 
+As a result, the runtime may not reclaim memory without delay, which can temporarily affect Eventing memory consumption.
+The service invokes a stop-the-world JavaScript runtime GC to reduce memory inside the JavaScript runtime isolate running each function.
```
```suggestion
The Eventing Service's garbage collection (GC) frees memory occupied by unreachable objects that are no longer in use by Eventing Functions.

Garbage collection reclaims memory only from objects that are no longer in use. 
Memory held by live objects in the JavaScript runtime heap remains allocated, so total usage may exceed the per-worker quota.

When memory usage approaches the per-worker quota limit, the JavaScript runtime may trigger garbage collection to reclaim memory. 

The garbage collector may delay the stop-the-world collection because it's optimized for throughput. 

The runtime might not reclaim memory without a delay, which can temporarily affect the Eventing Service's memory consumption.

The Eventing Service invokes a stop-the-world JavaScript runtime garbage collection to reduce memory inside the JavaScript runtime isolate running each Eventing Function.
```

sarahlwelton commented on modules/learn/pages/buckets-memory-and-storage/eventing-memory-quota.adoc:
```diff
@@ -0,0 +1,166 @@
+Live objects in JavaScript runtime heaps may persist, so memory usage can exceed the quota.
+* No OS-level enforcement outside cgroups: In VM or bare-metal deployments, the operating system does not enforce memory limits.
+* Worker sizing impacts memory: Configuring an excessive number of workers or running memory-intensive functions can increase memory pressure and affect node stability.
+* Quota applies per worker, not per function: Each worker receives an equal share of the total quota, regardless of individual function memory requirements.
```
```suggestion
== Limitations and Considerations

To prevent operational issues while using the Eventing Service, keep the following in mind while configuring the memory quota:

* <<no-hard-limit,>>
* <<no-enforce-c-groups,>>
* <<worker-sizing-memory,>>
* <<quota-per-worker,>>
```

sarahlwelton commented on modules/learn/pages/buckets-memory-and-storage/eventing-memory-quota.adoc:
```diff
@@ -0,0 +1,166 @@
+* Worker sizing impacts memory: Configuring an excessive number of workers or running memory-intensive functions can increase memory pressure and affect node stability.
+* Quota applies per worker, not per function: Each worker receives an equal share of the total quota, regardless of individual function memory requirements.
+
+=== Not a Hard Limit
```
```suggestion
[#no-hard-limit]
=== Not a Hard Limit
```

sarahlwelton commented on modules/learn/pages/buckets-memory-and-storage/eventing-memory-quota.adoc:
```diff
@@ -0,0 +1,166 @@
+* The quota does not account.
+* Continue to consume memory beyond the configured limit.
+
+As a result, setting a memory quota does not guarantee that Eventing functions remain within that limit.
```
```suggestion
* Garbage collection cannot reclaim.
* The quota does not include.
* Continue to consume memory beyond the configured limit.

As a result, setting a memory quota does not guarantee that Eventing Functions remain within that limit.
```

sarahlwelton commented on modules/learn/pages/buckets-memory-and-storage/eventing-memory-quota.adoc:
```diff
@@ -0,0 +1,166 @@
+
+As a result, setting a memory quota does not guarantee that Eventing functions remain within that limit.
+
+=== No OS-Level Enforcement Outside cgroups
```
```suggestion
[#no-enforce-c-groups]
=== No OS-Level Enforcement Outside cgroups
```

sarahlwelton commented on modules/learn/pages/buckets-memory-and-storage/eventing-memory-quota.adoc:
```diff
@@ -0,0 +1,166 @@
+
+=== No OS-Level Enforcement Outside cgroups
+
+In non-containerized environments, the operating system does not enforce the memory quota.
```
The operating system? Or Couchbase Server/the Eventing Service?

Pallavi-Janardhan replied:
Need to get this verified by the SME

Pallavi-Janardhan replied:
Spoke to the SME and have changed the title of this section, added new content, and modified this section.

sarahlwelton commented on modules/learn/pages/buckets-memory-and-storage/eventing-memory-quota.adoc:
```diff
@@ -0,0 +1,166 @@
+* Users must monitor Eventing memory usage using system tools.
+* Configuring an excessive number of workers can lead to memory exhaustion.
+* High-memory-consuming functions can affect overall system stability.
+* Eventing does not automatically stop or throttle execution when total memory exceeds the quota. 
```
```suggestion
You must monitor the Eventing Service's memory usage using system tools.

If you configure too many workers for your Eventing Functions, you can cause memory exhaustion in your cluster. 
High memory-consuming Eventing Functions can affect your overall system stability.

Again, the Eventing Service does not automatically stop or throttle execution when total memory exceeds the quota. 
```

sarahlwelton commented on modules/learn/pages/buckets-memory-and-storage/eventing-memory-quota.adoc:
```diff
@@ -0,0 +1,166 @@
+
+* Consider the per-worker quota when determining the number of workers.
+* Avoid over-provisioning workers when Eventing functions require substantial memory.
+* Balance the worker count against available system memory.
```
```suggestion
[#worker-sizing-memory]
=== Worker Sizing Impacts Memory

When configuring the memory quota for the Eventing Service, consider the per-worker quota when determining the number of workers.
Avoid over-provisioning workers when Eventing Functions require substantial memory.

For best results, try to balance your worker count against your available memory. 

[#quota-per-worker]
=== Quota Applies Per Worker

The memory quota does not isolate memory usage between Eventing Functions.
All workers draw from the same total memory quota.

A memory-intensive function can reduce the memory available to other Eventing Functions.

Eventing does not support per-function memory limits or reservations.

== Best Practices

Follow these recommendations to effectively manage Eventing memory usage: 

* <<monitor,>>
* <<test,>>
* <<use-cgroups,>>
* <<size,>>
* <<review,>>

[#monitor]
=== Monitor Memory Usage

* Use system-level monitoring tools to track actual memory consumption.
* Monitor memory usage for the Eventing process and individual workers.
* Configure alerts for memory thresholds well below system limits.
```

sarahlwelton commented on modules/learn/pages/buckets-memory-and-storage/eventing-memory-quota.adoc:
```diff
@@ -0,0 +1,166 @@
+* Optimize JavaScript code to minimize memory allocation.
+* Avoid accumulating large data structures in function scope.
+* Release object references when data is no longer needed.
+* Profile functions to identify memory-intensive operations.
```
```suggestion
[#test]
=== Test in Representative Environments

* Test Eventing Functions under production-like load conditions.
* Measure actual memory consumption during peak processing.
* Validate that memory usage remains within acceptable limits.

[#use-cgroups]
=== Use cgroups for Enforcement

* Consider containerized deployments or cgroup-based memory limits in production environments.
* Container platforms such as Docker and Kubernetes provide hard memory limits.
* cgroups enforce OS-level memory restrictions.
This approach adds a safety layer beyond the Eventing Service memory quota.

[#size]
=== Size the Memory Quota Appropriately

* Set the Eventing Service memory quota based on expected worker count and Eventing Function complexity.
* Allow additional headroom on top of the 256{nbsp}MB default for production workloads.
* Account for peak processing scenarios, not just average load.

[#review]
=== Review Function Memory Efficiency

* Optimize JavaScript code to minimize memory allocation.
* Avoid accumulating large data structures in the function scope.
* Release object references when data is no longer needed.
* Profile Eventing Functions to identify memory-intensive operations.
```

sarahlwelton reviewed (changes requested):
This needs a lot of work. Did you go through writing a doc plan for this work? I'm not sure that you're on the right track with how you chose to approach this, and we might have been able to guide you with a doc plan ahead of time.

sarahlwelton reviewed (changes requested):
This needs a lot of work. Did you go through writing a doc plan for this work? I'm not sure that you're on the right track with how you chose to approach this, and we might have been able to guide you with a doc plan ahead of time.

abhijpes commented:
Hi @Pallavi-Janardhan  Following up on the prior discussion, after further review I believe the following location is appropriate for a dedicated Eventing Memory Quota documentation page:

<img width="1709" height="735" alt="Screenshot 2026-01-08 at 2 04 16 PM" src="https://github.com/user-attachments/assets/0ae7260f-a83e-4464-bfe3-059bf9ce9d6a" />

simon-dew reviewed (commented):
As discussed with @Pallavi-Janardhan today, my comments are:

1. Remove `eventing-memory-quota.adoc` from this PR and revert the changes to the navigation. In the Sizing Guidelines document, keep a link to `xref:eventing:eventing-memory-quota.adoc` (see below) and any style changes, following Sarah's comments.
2. Create a new PR in the docs-devex repo to add `eventing-memory-quota.adoc` to the `eventing` module, as suggested by @abhijpes.

simon-dew commented on modules/install/pages/sizing-general.adoc:
```diff
@@ -554,68 +557,95 @@ The Analytics Service is dependent on the Data Service and requires the Data ser
 === RAM
 
-In general, the Eventing memory quota of 256 MB is sufficient for almost all workloads.
+The xref:learn:buckets-memory-and-storage/eventing-memory-quota.adoc[Eventing Memory Quota] doc discusses the Eventing Service memory quota in detail.
```
```suggestion
For more information about the Eventing Service memory quota, see xref:eventing:eventing-memory-quota.adoc[].
```

Pallavi-Janardhan replied:
Simon, I have provided the link but kept the wording as what Sarah suggested as I felt it was more apt. Hope that is fine!

simon-dew commented on modules/install/pages/sizing-general.adoc:
```diff
@@ -554,68 +557,95 @@ The Analytics Service is dependent on the Data Service and requires the Data ser
+* The distribution of the total memory quota across workers
+* The limitations and considerations
+* Best practices and recommendations
+--
```
Not sure if you need this bullet list? It's just a summary of`eventing-memory-quota.adoc`. I would delete this.

Pallavi-Janardhan replied:
Have deleted this list.

Pallavi-Janardhan committed:
Update modules/install/pages/sizing-general.adoc

Co-authored-by: sarahlwelton <110928505+sarahlwelton@users.noreply.github.com>

Pallavi-Janardhan committed:
Update modules/install/pages/sizing-general.adoc

Co-authored-by: sarahlwelton <110928505+sarahlwelton@users.noreply.github.com>

Pallavi-Janardhan committed:
Update modules/install/pages/sizing-general.adoc

Co-authored-by: sarahlwelton <110928505+sarahlwelton@users.noreply.github.com>

Pallavi-Janardhan committed:
Update modules/install/pages/sizing-general.adoc

Co-authored-by: sarahlwelton <110928505+sarahlwelton@users.noreply.github.com>

Pallavi-Janardhan committed:
Update modules/install/pages/sizing-general.adoc

Co-authored-by: sarahlwelton <110928505+sarahlwelton@users.noreply.github.com>

Pallavi-Janardhan committed:
Update modules/install/pages/sizing-general.adoc

Co-authored-by: sarahlwelton <110928505+sarahlwelton@users.noreply.github.com>

Pallavi-Janardhan committed:
Update modules/install/pages/sizing-general.adoc

Co-authored-by: sarahlwelton <110928505+sarahlwelton@users.noreply.github.com>

Pallavi-Janardhan committed:
Update modules/install/pages/sizing-general.adoc

Co-authored-by: sarahlwelton <110928505+sarahlwelton@users.noreply.github.com>

Pallavi-Janardhan committed:
adding the deleted heading RAM Sizing for Data Services Nodes

Pallavi-Janardhan committed:
Merge remote-tracking branch 'origin/DOC-13422-Eventing-Memory-Quota' into DOC-13422-Eventing-Memory-Quota

Pallavi-Janardhan committed:
Update modules/install/pages/sizing-general.adoc

Co-authored-by: sarahlwelton <110928505+sarahlwelton@users.noreply.github.com>

Pallavi-Janardhan committed:
Update modules/install/pages/sizing-general.adoc

Co-authored-by: sarahlwelton <110928505+sarahlwelton@users.noreply.github.com>

Pallavi-Janardhan committed:
Update modules/install/pages/sizing-general.adoc

Co-authored-by: sarahlwelton <110928505+sarahlwelton@users.noreply.github.com>

Pallavi-Janardhan committed:
Update modules/install/pages/sizing-general.adoc

Co-authored-by: sarahlwelton <110928505+sarahlwelton@users.noreply.github.com>

Pallavi-Janardhan committed:
Update modules/install/pages/sizing-general.adoc

Co-authored-by: sarahlwelton <110928505+sarahlwelton@users.noreply.github.com>

Pallavi-Janardhan committed:
Update modules/install/pages/sizing-general.adoc

Co-authored-by: sarahlwelton <110928505+sarahlwelton@users.noreply.github.com>

Pallavi-Janardhan committed:
Update modules/install/pages/sizing-general.adoc

Co-authored-by: sarahlwelton <110928505+sarahlwelton@users.noreply.github.com>

Pallavi-Janardhan committed:
Update modules/install/pages/sizing-general.adoc

Co-authored-by: sarahlwelton <110928505+sarahlwelton@users.noreply.github.com>

Pallavi-Janardhan committed:
Update modules/install/pages/sizing-general.adoc

Co-authored-by: sarahlwelton <110928505+sarahlwelton@users.noreply.github.com>

Pallavi-Janardhan committed:
Update modules/install/pages/sizing-general.adoc

Co-authored-by: sarahlwelton <110928505+sarahlwelton@users.noreply.github.com>

Pallavi-Janardhan committed:
Update modules/install/pages/sizing-general.adoc

Co-authored-by: sarahlwelton <110928505+sarahlwelton@users.noreply.github.com>

Pallavi-Janardhan committed:
Update modules/install/pages/sizing-general.adoc

Co-authored-by: sarahlwelton <110928505+sarahlwelton@users.noreply.github.com>

Pallavi-Janardhan committed:
Update modules/install/pages/sizing-general.adoc

Co-authored-by: sarahlwelton <110928505+sarahlwelton@users.noreply.github.com>

Pallavi-Janardhan committed:
Update modules/install/pages/sizing-general.adoc

Co-authored-by: sarahlwelton <110928505+sarahlwelton@users.noreply.github.com>

Pallavi-Janardhan committed:
Update modules/install/pages/sizing-general.adoc

Co-authored-by: sarahlwelton <110928505+sarahlwelton@users.noreply.github.com>

Pallavi-Janardhan committed:
Update modules/install/pages/sizing-general.adoc

Co-authored-by: sarahlwelton <110928505+sarahlwelton@users.noreply.github.com>

Pallavi-Janardhan committed:
Update modules/install/pages/sizing-general.adoc

Co-authored-by: sarahlwelton <110928505+sarahlwelton@users.noreply.github.com>

Pallavi-Janardhan committed:
Update modules/install/pages/sizing-general.adoc

Co-authored-by: sarahlwelton <110928505+sarahlwelton@users.noreply.github.com>

Pallavi-Janardhan committed:
Update modules/install/pages/sizing-general.adoc

Co-authored-by: sarahlwelton <110928505+sarahlwelton@users.noreply.github.com>

Pallavi-Janardhan committed:
Update modules/install/pages/sizing-general.adoc

Co-authored-by: sarahlwelton <110928505+sarahlwelton@users.noreply.github.com>

Pallavi-Janardhan committed:
Update modules/install/pages/sizing-general.adoc

Co-authored-by: sarahlwelton <110928505+sarahlwelton@users.noreply.github.com>

Pallavi-Janardhan committed:
Update modules/install/pages/sizing-general.adoc

Co-authored-by: sarahlwelton <110928505+sarahlwelton@users.noreply.github.com>

Pallavi-Janardhan committed:
Update modules/install/pages/sizing-general.adoc

Co-authored-by: sarahlwelton <110928505+sarahlwelton@users.noreply.github.com>

Pallavi-Janardhan committed:
Update modules/install/pages/sizing-general.adoc

Co-authored-by: sarahlwelton <110928505+sarahlwelton@users.noreply.github.com>

Pallavi-Janardhan committed:
Update modules/install/pages/sizing-general.adoc

Co-authored-by: sarahlwelton <110928505+sarahlwelton@users.noreply.github.com>

Pallavi-Janardhan committed:
Update modules/install/pages/sizing-general.adoc

Co-authored-by: sarahlwelton <110928505+sarahlwelton@users.noreply.github.com>

Pallavi-Janardhan committed:
Update modules/install/pages/sizing-general.adoc

Co-authored-by: sarahlwelton <110928505+sarahlwelton@users.noreply.github.com>

Pallavi-Janardhan committed:
Update modules/install/pages/sizing-general.adoc

Co-authored-by: sarahlwelton <110928505+sarahlwelton@users.noreply.github.com>

Pallavi-Janardhan committed:
Update modules/install/pages/sizing-general.adoc

Co-authored-by: sarahlwelton <110928505+sarahlwelton@users.noreply.github.com>

Pallavi-Janardhan committed:
Update modules/install/pages/sizing-general.adoc

Co-authored-by: sarahlwelton <110928505+sarahlwelton@users.noreply.github.com>

Pallavi-Janardhan committed:
incorporating editorial review comments

Pallavi-Janardhan committed:
Update modules/install/pages/sizing-general.adoc

Co-authored-by: sarahlwelton <110928505+sarahlwelton@users.noreply.github.com>

Pallavi-Janardhan committed:
Update modules/learn/pages/buckets-memory-and-storage/eventing-memory-quota.adoc

Co-authored-by: sarahlwelton <110928505+sarahlwelton@users.noreply.github.com>

Pallavi-Janardhan committed:
Update modules/learn/pages/buckets-memory-and-storage/eventing-memory-quota.adoc

Co-authored-by: sarahlwelton <110928505+sarahlwelton@users.noreply.github.com>

Pallavi-Janardhan committed:
Update modules/learn/pages/buckets-memory-and-storage/eventing-memory-quota.adoc

Co-authored-by: sarahlwelton <110928505+sarahlwelton@users.noreply.github.com>

Pallavi-Janardhan committed:
Update modules/learn/pages/buckets-memory-and-storage/eventing-memory-quota.adoc

Co-authored-by: sarahlwelton <110928505+sarahlwelton@users.noreply.github.com>

Pallavi-Janardhan committed:
Update modules/learn/pages/buckets-memory-and-storage/eventing-memory-quota.adoc

Co-authored-by: sarahlwelton <110928505+sarahlwelton@users.noreply.github.com>

Pallavi-Janardhan committed:
Update modules/learn/pages/buckets-memory-and-storage/eventing-memory-quota.adoc

Co-authored-by: sarahlwelton <110928505+sarahlwelton@users.noreply.github.com>

Pallavi-Janardhan committed:
Update modules/learn/pages/buckets-memory-and-storage/eventing-memory-quota.adoc

Co-authored-by: sarahlwelton <110928505+sarahlwelton@users.noreply.github.com>

Pallavi-Janardhan committed:
Update modules/learn/pages/buckets-memory-and-storage/eventing-memory-quota.adoc

Co-authored-by: sarahlwelton <110928505+sarahlwelton@users.noreply.github.com>

Pallavi-Janardhan committed:
Update modules/learn/pages/buckets-memory-and-storage/eventing-memory-quota.adoc

Co-authored-by: sarahlwelton <110928505+sarahlwelton@users.noreply.github.com>

Pallavi-Janardhan committed:
Update modules/learn/pages/buckets-memory-and-storage/eventing-memory-quota.adoc

Co-authored-by: sarahlwelton <110928505+sarahlwelton@users.noreply.github.com>

Pallavi-Janardhan committed:
Update modules/learn/pages/buckets-memory-and-storage/eventing-memory-quota.adoc

Co-authored-by: sarahlwelton <110928505+sarahlwelton@users.noreply.github.com>

Pallavi-Janardhan committed:
Update modules/learn/pages/buckets-memory-and-storage/eventing-memory-quota.adoc

Co-authored-by: sarahlwelton <110928505+sarahlwelton@users.noreply.github.com>

Pallavi-Janardhan committed:
Update modules/learn/pages/buckets-memory-and-storage/eventing-memory-quota.adoc

Co-authored-by: sarahlwelton <110928505+sarahlwelton@users.noreply.github.com>

Pallavi-Janardhan committed:
Update modules/learn/pages/buckets-memory-and-storage/eventing-memory-quota.adoc

Co-authored-by: sarahlwelton <110928505+sarahlwelton@users.noreply.github.com>

Pallavi-Janardhan committed:
Update modules/learn/pages/buckets-memory-and-storage/eventing-memory-quota.adoc

Co-authored-by: sarahlwelton <110928505+sarahlwelton@users.noreply.github.com>

Pallavi-Janardhan committed:
Update modules/learn/pages/buckets-memory-and-storage/eventing-memory-quota.adoc

Co-authored-by: sarahlwelton <110928505+sarahlwelton@users.noreply.github.com>

Pallavi-Janardhan committed:
Update modules/learn/pages/buckets-memory-and-storage/eventing-memory-quota.adoc

Co-authored-by: sarahlwelton <110928505+sarahlwelton@users.noreply.github.com>

Pallavi-Janardhan committed:
Update modules/learn/pages/buckets-memory-and-storage/eventing-memory-quota.adoc

Co-authored-by: sarahlwelton <110928505+sarahlwelton@users.noreply.github.com>

Pallavi-Janardhan committed:
Update modules/learn/pages/buckets-memory-and-storage/eventing-memory-quota.adoc

Co-authored-by: sarahlwelton <110928505+sarahlwelton@users.noreply.github.com>

Pallavi-Janardhan committed:
Update modules/learn/pages/buckets-memory-and-storage/eventing-memory-quota.adoc

Co-authored-by: sarahlwelton <110928505+sarahlwelton@users.noreply.github.com>

Pallavi-Janardhan committed:
Update modules/learn/pages/buckets-memory-and-storage/eventing-memory-quota.adoc

Co-authored-by: sarahlwelton <110928505+sarahlwelton@users.noreply.github.com>

Pallavi-Janardhan committed:
Update modules/learn/pages/buckets-memory-and-storage/eventing-memory-quota.adoc

Co-authored-by: sarahlwelton <110928505+sarahlwelton@users.noreply.github.com>

Pallavi-Janardhan committed:
Update modules/learn/pages/buckets-memory-and-storage/eventing-memory-quota.adoc

Co-authored-by: sarahlwelton <110928505+sarahlwelton@users.noreply.github.com>

Pallavi-Janardhan committed:
Update modules/learn/pages/buckets-memory-and-storage/eventing-memory-quota.adoc

Co-authored-by: sarahlwelton <110928505+sarahlwelton@users.noreply.github.com>

Pallavi-Janardhan committed:
Update modules/learn/pages/buckets-memory-and-storage/eventing-memory-quota.adoc

Co-authored-by: sarahlwelton <110928505+sarahlwelton@users.noreply.github.com>

Pallavi-Janardhan committed:
Update modules/learn/pages/buckets-memory-and-storage/eventing-memory-quota.adoc

Co-authored-by: sarahlwelton <110928505+sarahlwelton@users.noreply.github.com>

Pallavi-Janardhan committed:
Update modules/learn/pages/buckets-memory-and-storage/eventing-memory-quota.adoc

Co-authored-by: sarahlwelton <110928505+sarahlwelton@users.noreply.github.com>

Pallavi-Janardhan committed:
Update modules/learn/pages/buckets-memory-and-storage/eventing-memory-quota.adoc

Co-authored-by: sarahlwelton <110928505+sarahlwelton@users.noreply.github.com>

Pallavi-Janardhan committed:
incorporating editorial review comments

Pallavi-Janardhan committed:
adding path to the Eventing memory quota file in eventing directory

Pallavi-Janardhan committed:
fixing typo and link in the sizing-genera file

Pallavi-Janardhan committed:
deleting eventing-memory-quota.adoc file from the docs-server repo

Pallavi-Janardhan committed:
deleting the link to the eventing-memory-quota.adoc file from the nav.adoc file in the docs-server repo

Pallavi-Janardhan committed:
gave a fresh link to the eventing-memory-quota.adoc file from the nav.adoc file in the docs-server repo

osfameron committed:
DOC-13422 add preview config

This will also build the related docs-devex branch

Pallavi-Janardhan commented:
As discussed with Abhishek Jindal, I have added a section --> "Memory Management in Containerized vs Non-Containerized Environments", in the Eventing Memory Quota file. 

As discussed with my doc architects, am creating another PR just for the Eventing file that is going into another repo - **docs-devex** repo. 

The PR for that is --> https://github.com/couchbaselabs/docs-devex/pull/555

Please do review both the PRs and let me know your comments/approval. 

Thanks and regards,
Pallavi

Pallavi-Janardhan committed:
deleted some bulletted reference points that were suggested to be deleted

sarahlwelton commented on modules/install/pages/sizing-general.adoc:
```diff
@@ -507,7 +515,8 @@ Based on the above formula, these are the suggested sizing guidelines:
 |===
 
-The above example shows the memory requirement of a secondary index with 10M index entries, each with 50 bytes size of secondary key and 30 bytes size of documentID. The memory usage requirements are 2.5GB(Nitro, 100% resident), 1GB(plasma, 20% resident), 800MB(Forestdb, 20% resident).
+The above example shows the memory requirement of a secondary index with 10M index entries, each with 50 bytes size of secondary key and 30 bytes size of documentID. 
```
```suggestion
The previous example shows the memory requirement of a secondary index with 10M index entries, each with a 50 bytes secondary key and a 30 bytes DocumentID. 
```

sarahlwelton commented on modules/install/pages/sizing-general.adoc:
```diff
@@ -554,68 +563,87 @@ The Analytics Service is dependent on the Data Service and requires the Data ser
 For example, if you process 100K mutations per second and only match 1 out of 1000 patterns, then perform some intense computation on the matched 100 items in your Eventing Function, you need 100X less compute than if you performed the intense computation on each mutation.
 
-Eventing also can perform I/O to external REST endpoints via a synchronous HTTP/S cURL call. In this case, Eventing typically blocks on I/O and doesn’t need much CPU. However. if you want high throughput to overcome bandwidth, you will need more workers and thus more cores.
+Eventing also can perform I/O to external REST endpoints via a synchronous HTTP/S cURL call. 
```
```suggestion
Eventing also can perform I/O to external REST endpoints through a synchronous HTTP/S cURL call. 
```

sarahlwelton commented on modules/install/pages/sizing-general.adoc:
```diff
@@ -554,68 +563,87 @@ The Analytics Service is dependent on the Data Service and requires the Data ser
 === RAM
 
-In general, the Eventing memory quota of 256 MB is sufficient for almost all workloads.
+The xref:eventing:eventing-memory-quota.adoc[Eventing Memory Quota] doc discusses the Eventing Service Memory quota in detail.
```
```suggestion
For more information about how to size your Eventing memory quota, see xref:eventing:eventing-memory-quota.adoc[].
```

sarahlwelton commented on modules/install/pages/sizing-general.adoc:
```diff
@@ -554,68 +563,87 @@ The Analytics Service is dependent on the Data Service and requires the Data ser
+100,000 active timers require 100 MB of additional bucket space.
 
-As a best practice, it's recommended to keep this collection 100% resident, so that it's always available in-memory.
+As a best practice, keep this collection fully resident in-memory to make sure constant availability.
```
```suggestion
As a best practice, keep this collection fully resident in-memory to make sure you have constant availability.
```

sarahlwelton commented on modules/install/pages/sizing-general.adoc:
```diff
@@ -631,13 +659,14 @@ The hardware requirements for running a backup cluster are as follows:
-For example, XDCR needs 1-2 additional CPU cores per stream. In some cases, it also requires additional RAM and network resources. If a cluster is not sized to handle _both_ the existing workload _and_ the new XDCR streams, the performance of both XDCR and the cluster overall may be negatively impacted.
+For example, XDCR needs 1-2 additional CPU cores per stream. 
+In some cases, it also requires additional RAM and network resources. 
+If a cluster is not sized to handle `both` the existing workload `and` the new XDCR streams, the performance of both XDCR and the cluster overall may be negatively impacted.
```
```suggestion
If a cluster is not sized to handle both the existing workload and the new XDCR streams, the performance of both XDCR and the cluster overall might be negatively impacted.
```

sarahlwelton reviewed (changes requested):
Still some issues. Thanks.

Pallavi-Janardhan committed:
Update modules/install/pages/sizing-general.adoc

Co-authored-by: sarahlwelton <110928505+sarahlwelton@users.noreply.github.com>

Pallavi-Janardhan committed:
Update modules/install/pages/sizing-general.adoc

Co-authored-by: sarahlwelton <110928505+sarahlwelton@users.noreply.github.com>

Pallavi-Janardhan committed:
Update modules/install/pages/sizing-general.adoc

Co-authored-by: sarahlwelton <110928505+sarahlwelton@users.noreply.github.com>

Pallavi-Janardhan committed:
Apply suggestions from code review

Making few editorial review comments

Co-authored-by: sarahlwelton <110928505+sarahlwelton@users.noreply.github.com>

Pallavi-Janardhan committed:
Update modules/install/pages/sizing-general.adoc

Co-authored-by: sarahlwelton <110928505+sarahlwelton@users.noreply.github.com>

Pallavi-Janardhan commented:
Have made all the suggested modifications. Please review and approve.

sarahlwelton reviewed (commented):
One comment.

Pallavi-Janardhan committed:
Update modules/install/pages/sizing-general.adoc

Co-authored-by: sarahlwelton <110928505+sarahlwelton@users.noreply.github.com>

Pallavi-Janardhan commented:
Made all necessary edits.

sarahlwelton reviewed (approved):
I'm sure we could still make more improvements but this is fine for now.

sarahlwelton reviewed (approved):
I already approved and then you wiped out my approval by re-requesting review @Pallavi-Janardhan . Please be careful in the future.

simon-dew commented on preview/DOC-13422-Eventing-Memory-Quota.yml:
Don't forget to delete this file before merging the PR, so the `preview` folder does not get cluttered with old preview config files

Pallavi-Janardhan replied:
Deleted, Simon. Now merging.

simon-dew reviewed (approved):
LGTM 👍

Pallavi-Janardhan committed:
Delete preview/DOC-13422-Eventing-Memory-Quota.yml

Deleting the yml file as suggested so the preview folder does not get cluttered with old preview config files
